### DISCUSSION TITLE
Vulkan API backend

### DISCUSSION THREAD

**Discussion by Appricorn4** [Comfy-Org/ComfyUI#5530] [Ideas]
Is a Vulkan backend possible? I use Vulkan on stable-diffusion.cpp and it's significantly faster than CLBlast, CUDA, and ROCm in my testing. I would really like to see it here as Vulkan support is universal and easy, and would help especially with AMD and Intel GPUs (and possibly Qualcomm).

**Comment by alshdavid** [comment]
+1 this should be a priority as it unlocks the use of virtually every device. Consider;
- AMD's "AI Max" APUs 
  - Can have 100gb+ video RAM
  - Currently no support for ROCm  
- Qualcomm/Rockchip/etc chips 
- AMD 9000 series cards 
  - Currently no (practical) support for ROCm 
- Potentially even MacOS devices via MoltenVK.

**Comment by chew1976** [comment]
Please make Vulkan possible. I bought a Strix Halo 128gb just to use Comfy UI, only to find out that I am dying to use Comfy UI on it.

**Comment by alshdavid** [comment]
CC https://github.com/pytorch/pytorch/issues/160230

**Comment by kamathln** [comment]
Using Vulkan made installing llama a breeze on my Linux distro that has trouble with ROCm. Wish every AI tool has fist class support for Vulkan first and then targets specific tools.

**Comment by dc3000** [comment]
Reply +1

**Comment by DocMAX** [comment]
+1

**Comment by fgheorghe** [comment]
+1 Vulkan would be awesome.

**Comment by Fugach** [comment]
+1 amazing idea

**Comment by Jhovala** [comment]
+1

**Comment by nololler** [comment]
+1

**Comment by jark006** [comment]
+1

**Comment by Tamalero** [comment]
+1

**Comment by darkman7hr** [comment]
+1

**Comment by teizumika** [comment]
+1

**Comment by wychuzord** [comment]
+1 Practically I can't use AI without Vulkan on my machine

**Comment by hzr12** [comment]
+1

**Comment by NoThanksGoAway** [comment]
Wouldn't that be on pytorch's side and not really specific to Comfy? Or am I insane? Rocm sucks though and I'm tired of pretending it doesn't. =)

**Reply by alshdavid** [reply]
For sure, ticket for it here: https://github.com/pytorch/pytorch/issues/160230

**Reply by wychuzord** [reply]
Llama.cpp has Vulkan capability without need for pytorch (or at least I didn't need to bother with pytorch flatpack LM Studio?).

**Reply by alshdavid** [reply]
I believe Llama.cpp cannot be used for diffusion models, there is [stable-diffusion.cpp](https://github.com/leejet/stable-diffusion.cpp#stable-diffusioncpp) however - which is based on the same underlying libraries and supports Vulkan.

My guess is that it's impractical to migrate from pytorch

**Reply by alshdavid** [reply]
CC: https://github.com/Comfy-Org/ComfyUI/discussions/5530#discussioncomment-17687154

**Comment by NoThanksGoAway** [comment]
Sd.cpp's vulkan backend works really well.

**Reply by alshdavid** [reply]
CC: https://github.com/Comfy-Org/ComfyUI/discussions/5530#discussioncomment-17687154

**Comment by wenyhtt** [comment]
+1

**Comment by JohnD87** [comment]
+1

**Comment by GreatestTrain** [comment]
+1

**Comment by alshdavid** [comment]
I was curious about sd.cpp so I experimented with a reference implementation to test out the performance compared to ROCm. 
Wrote a simple CLI tool that lets you run diffusion with SDXL based models.

You can download it here: [https://github.com/alshdavid/diffusion-vk/releases/tag/0.0.1](https://github.com/alshdavid/diffusion-vk/releases/tag/0.0.1)
Source code is here: [https://github.com/alshdavid/diffusion-vk](https://github.com/alshdavid/diffusion-vk)

It's unoptimised, no attention optimisations but even still it's sometimes faster than ROCm via ComfyUI, and always consistent

- Windows 10
- 9070xt 16gb ROCm
- Image Size: 1024 x 768
- Steps 15

|||
|-|-|
| ROCm ComfyUI | 10s - 26s |
| sd.cpp | 11s (consistently) |

Why does AMD even bother with ROCm/HIP

**Reply by OrsoEric** [reply]
The generative text node really exemplifies the enormous difference in performance 

[my 7900XTX gets 14TPS on Qwen 3 VL 4B for Krea2, but gets 90TPS for Qwen 3.5 9B Q4 on llama.cpp vulkan.](https://github.com/OrsoEric/HOWTO-ComfyUI#llm-imagetext-safetensor)

**Reply by dduede** [reply]
@OrsoEric Is this even a fair comparison? RDNA3 doesnt support fp8 natively and needs to fallback to a slower compute path, which inherently makes it slower than a Q4 quantized model. You are also comparing two different models and model sizes. I'm all for getting more light on vulkan and pushing it further - but this comparison of yours is harmful. Llamacpps vulkan implementation will most likely be faster than pytorches rocm backend, but your statement makes it look worse than it really is.

From personal experience, I can say that vulkan overall seems to be 10-15% faster on average compared to ROCm but there are scenarios where ROCm is faster. Generally, ROCm should be faster than Vulkan, but optimizations are tricky. If a contributor pushes out an update to ROCm in pytorch that suddenly boosts its performance past anything we see with any vulkan implementation, a lot of people advocating for vulkan would probably just silently switch back to ROCm.

I want to see pytorch vulkan support not because "its faster right now", but because it potentially opens the doors for heterogenic setups that use different GPU types in parallel in the future.

**Reply by OrsoEric** [reply]
> Is this even a fair comparison?

Yes.

I just tested on my 358H laptop, and Qwen 3 4B VL Q4 pulls 21.68 tokens per second with vulkan backend. A laptop with iGPU obliterates a ROCm with a 7900XTX by 50%. It's an humiliating difference. That's the gulf between high performance Vulkan, vs ROCm under windows.

> I want to see pytorch vulkan support not because "its faster right now"

1) ROCm on windows is an aftertought. We got the first dlls THIS YEAR. It's linux first, meaning it already cuts 90% of the user base. It's just unfit for duty.

2) Vulkan is (vastly) superior to ROCm because the AMD's Adrenaline team is working on it, and if the Adrenaline team isn't firing on all cylinders, AMD sells no card at all. AMD was forced to make good Vulkan, because without it might as well close the Radeon division. Why do you think AMD skipped cards and generations in the past? Software held the cards back, even when the hardware team had decent compute pipelines. And AMD has TWO independent software pipelines? WHY? Seriously: WHY?

Make pytorch/Vulkan! You already have good people working on the Vulkan/AMD pathways! Don't reinvent the wheels to do pytorch/ROCm/AMD!

Until AMD takes ROCm under windows seriously, it's not even a comparison. It's dumb luck when a workload is accelerated under ROCm.

I got my 358H with B390. Zimage on 7900XTX pulls 16 seconds, my laptop pulls 50s. That' humiliating for ROCm, the 7900XTX has 9.6X the bandwidth, and 8.5X the raw FP16 FLOPS. And it pulls barely 3X performance. The performance loss of ROCm is inexcusable. And that's not against an high performance CUDA backend, but against they pytorch Intel backend! Intel made better pytorch bindings!!!

Vulkan is the only way to get the 7900XTX hardware to shine and put its compute and bandwidth to task.

**Comment by hwertz** [comment]
Just wanted to also throw in, at least on Ubuntu with up to date Mesa, LM Studio (llama) Vulkan support even works on something like a Coffee Lake integrated GPU.  Not sure that it actually is faster than just running on GPU; but nevertheless, it did run correctly, and there's no arbitrary VRAM limit, on my 32GB system I ran one that was using like 20GB VRAM.

So this would also presumably get all but the oldest Intel GPUs going.  I did play with this on a Tiger Lake and the GPU on that is at least faster than doing it on the CPU.

**Comment by vlforit** [comment]
So we can start from this point to bring Vulkan for ComfyUI

```
diff --git a/comfy/model_management.py b/comfy/model_management.py
index ff963eb8..6013b2e6 100644
--- a/comfy/model_management.py
+++ b/comfy/model_management.py
@@ -194,6 +194,8 @@ def is_wsl():
 def get_torch_device():
     global directml_enabled
     global cpu_state
+    if torch.is_vulkan_available():
+        return torch.device(device='vulkan')
     if directml_enabled:
         global directml_device
         return directml_device
@@ -321,7 +323,10 @@ def get_total_memory(dev=None, torch_total_too=False):
         mem_total = psutil.virtual_memory().total
         mem_total_torch = mem_total
     else:
-        if directml_enabled:
+        if torch.is_vulkan_available():
+            mem_total = 1024 * 1024 * 1024 * 24 #24Gb 7900 xtx,  hardcoded
+            mem_total_torch = mem_total
+        elif directml_enabled:
             mem_total = 1024 * 1024 * 1024 #TODO
             mem_total_torch = mem_total
         elif is_intel_xpu():
```
